### DISCUSSION TITLE
Are there additional API calls?

### DISCUSSION THREAD

**Discussion by zramsis** [openmpf/openmpf#1573] [Q&A] [answered]
Hello!

My team and I have been working with OpenMPF for a couple months now and have been building an app that will run alongside it. We have been using your base api (the functions listed on Swagger) but also functions that are not, such as uploadFile. Until recently this has worked for us, we were able to use that api call to upload a file but as of two days ago we have been getting a 403 Forbidden in response to that call. We are not sure if an update was made recently/if this functionality was intended. Is there a way that those additional functions can be added to Swagger/could be used outside of the workflow manager? Another function we were hoping to be able to use was jobs-paged. 

We are currently running our app locally on 8081 while OpenMPF is running on 8080 so we've had to use a proxy to avoid cors errors. Is there a better way to avoid cors errors?

In general, functionality we are hoping to use with your api is:
get multiple jobs
upload file/files
delete file

Thanks,
Zak

**Comment by brosenberg42** [comment, marked as answer]
To enable CORS, you can set the `CORS_ALLOWED_ORIGINS` environment variable on Workflow Manager. It should be set to either the single origin where your web app is running or a comma-separated list of origins you would like to enable CORS for. 

The endpoints you mentioned are not part of the public API. The public API is documented on the Swagger page accessible through Workflow Manager. The public API endpoints all start with `workflow-manager/rest/`. We intentionally limit which endpoints are part of our public API because those are the endpoints that we are interesting in guaranteeing API stability for. We use semantic versioning to indicate compatibility breaking changes with the public REST API, but there is no such guarantee for the internal REST endpoints.

I think the reason the internal endpoint stopped working is that in OpenMPF version 7, we upgraded the version of Spring Security. The new version of Spring Security enables CSRF protection by default. I think the easiest thing for you to do is to use OpenMPF version 6.3.15. 

Another option is to give your other app access to the shared data volume. Then, your app can manage the files without interacting with Workflow Manager.

If you are interested in building your own version of Workflow Manager, you can add things to the public API. To add something to the public REST API you will need change the `@RequestMapping(value = "/some/endpoint")` annotation on the desired controller method to `@RequestMapping(value = {"/some/endpoint", "/rest/some/endpoint"})`. For your specific example you would need to change: https://github.com/openmpf/openmpf/blob/develop/trunk/workflow-manager/src/main/java/org/mitre/mpf/mvc/controller/MediaController.java#L180


Another option, if you are willing to build your own version of Workflow Manager, is to disable CSRF protection. You can do this by adding `<security:csrf disabled="true" />` here: https://github.com/openmpf/openmpf/blob/develop/trunk/workflow-manager/src/main/resources/applicationContext-spring-security.xml#L56

Another way to address the issue is to provide an HTTP URL in the `mediaUri` field. In that case, Workflow Manager will download the media from that URL and delete its local copy when the job completes. The Workflow Manager also supports [downloading the media from an S3 compatible server.](https://openmpf.github.io/docs/site/Object-Storage-Guide/index.html#s3-object-storage) Setting up a [MinIO server using their Docker image](https://hub.docker.com/r/minio/minio) is relatively easy.
